Flight modes dev guide updates

hamishwillee opened the pull request:
This updates the flight mode "dev" guide - https://docs.px4.io/main/en/concept/flight_modes.html

- Removes the diagram
- Removes info about specific modes - links to other docs.
- Adds info about Standard modes and mavlink integration

Lots still to do. This is draft.

Fixes #3564

hamishwillee commented on en/concept/flight_modes.md:
```diff
@@ -1,111 +1,130 @@
+## MAVLink Integration
 
-- **Rovers / Boats:**
+### Standard Modes Protocol
```
@bkueng This section is underway.

Some questions you may choose to answer, though it is early for a review.

- I can see that when you add a new mode, you don't need anything special for the current mode to update. 
- Do you need to do anything if you add a new mode to trigger the mode monitor?
- I recall for ROS2 registration will add a mode to the list of available modes. Do you need to do anything in P4 to add a new mode to the list of available modes?

hamishwillee commented on en/concept/flight_modes.md:
```diff
@@ -1,111 +1,215 @@
+
+[PX4 ROS 2 Control Interface](../ros2/px4_ros2_control_interface.md) documents how to create flight modes running on a companion computer.
 
+## PX Flight Modes (FC)
+
+The specific control behaviour of a flight mode at any time is determined by a [Flight Task](../concept/flight_tasks.md).
+A flight mode might define one or more tasks that define variations of the mode behavior.
+The actual behaviour that is used is normally defined in a parameter, and selected in [src/modules/flight_mode_manager/FlightModeManager.cpp](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/flight_mode_manager/FlightModeManager.cpp#L266-L285)
+
```
@MaEtUgR Can I get some hints for flight modes. It looks like the bit of the code that does something is the Flight Task, and that a mode might have a few of these all that do a particular variant of the mode behaviour - e.g. different types of position control.

I also understand the mode restrictions, which is in the next section (I think).

But not clear 
- how I declare an entirely new mode "HamishMode".
- Where I have to add the code for that.
- Whether I have to add that to the list of modes exported to MAVLink or that happens magically
- If there are any base classes - i.e. What is a mode.

Any chance of some bullet points of the key concepts?

hamishwillee commented on en/concept/flight_modes.md:
```diff
@@ -1,111 +1,215 @@
-## Flight Mode Summary
+## MAVLink Integration
+
+### Standard Modes Protocol
```
FYI I might move this whole section under https://docs.px4.io/main/en/middleware/mavlink.html

I do plan to:
- [ ] Add a mavlink section to the mode docs explaining how you can engage the standard modes. Custom modes.

hamishwillee commented on en/concept/flight_modes.md:
```diff
@@ -1,111 +1,215 @@
+At time of writing there are no PX4 ROS 2 flight modes that are considered "supported as an integral part of the PX4 open source offering".
+:::
+
+## PX4 FC Modes vs PX4 ROS2 Flight Modes
```
FYI, this section put me in contortions - how do I refer to a flight mode that runs in the FC?

It was the reason I wrote this https://github.com/PX4/PX4-user_guide/issues/3565#issuecomment-2655407569

sfuhrer replied:
Should we label them with "internal" (vs. "external"=running on ROS2)

mrpollo replied:
I like Internal vs External @sfuhrer

hamishwillee replied:
@mrpollo @sfuhrer I "like" Internal vs External too, but I'm not certain it is future proof terminology. 

Assuming you read https://github.com/PX4/PX4-user_guide/issues/3565#issuecomment-2655407569, are you suggesting that if in future we have a flight stack that spans FC and Companion, then we will refer to the parts as _PX4 Internal_ (runs on FC) and _PX4 External_ (runs on companion)?

It's a little clunky in that it will make any code on the companion a second class citizen. That might be OK. 
I proposed PX4 RTOS modes / PX4 GTOS modes, but if you are both happy with internal external for this, we can give it a go.

hamishwillee replied:
For now I will try Internal vs External, but I would like you to think about ^^^^

sfuhrer replied:
> "Internal" implies "you've got to be on the flight stack to really be part of PX4".

But that's how it is - we (PX4 community) do not maintain the ROS flight modes, so the user is on it's own when they use them.

I btw fully support you in spending the time now to think this through - it will be very hard to change in the future once people get used to the naming.

hamishwillee replied:
Yes, that is exactly the problem.

If I knew we'd have "full stack" PX4 in a year or so I'd push for `PX4 RTOS modes` and `PX4 GCOS modes` (or replacing RTOS/GCOS with RT, NuttX, GC, Linux).

sfuhrer reviewed (commented):
Looks concise. I unfortunately don't have enough exposure to external flight modes yet to highlight common pitfalls that we should cover with the docs.
One thing I find a bit unfortunate with the new structure (and with the Standard Modes) is that it's hard to distinguish "auto" from "manual" modes.

mrpollo commented on en/concept/flight_modes.md:
```diff
@@ -1,111 +1,215 @@
+- [Basic Configuration > Flight Modes](../config/flight_mode.md)
+
+::: info
+At time of writing there are no PX4 ROS 2 flight modes that are considered "supported as an integral part of the PX4 open source offering".
```
I wonder if this is needed, I don't think we will ever get to the point where there's an external flight mode that is required for PX4 to run. What's the intention here?

hamishwillee replied:
There have been statements made at various points that PX4 will eventually offer a software stack that spans the companion computer. That doesn't mean "required to run" so much as "officially supported software".

We're not there yet, but since we are moving towards an architecture in which companion is an expected part of the drone, isn't that the direction you wish to take?

Upshot, perhaps no one would assume there might be. But I'd like to understand where we are going.

mrpollo commented on en/concept/flight_modes.md:
```diff
@@ -1,111 +1,215 @@
+
+## PX4 FC Modes vs PX4 ROS2 Flight Modes
+
+With some exceptions it is possible to write a mode in either the FC and companion computer.
```
Maybe a good time to introduce the Internal vs External concept, and use that troughout the doc

hamishwillee replied:
Yes, as long as we agree the terminology - https://github.com/PX4/PX4-user_guide/pull/3577#discussion_r1960875336

mrpollo commented on en/concept/flight_modes.md:
```diff
@@ -1,111 +1,215 @@
+They have the following benefits:
+
+- Easier to implement as there is no need to deal with low-level embedded constraints and requirements (such as restricted stack sizes).
+- Easier to maintain as the integration API is small, well defined, and stable (porting custom PX4 modes on the flight controller between PX4 versions is much harder).
```
Maybe add a note separate from this bullet point with the warning under parenthesis, and extend on the reasons

hamishwillee replied:
@mrpollo My _assumption_ here is that PX4 dev team makes little effort in any respect support custom forks, and may well break any dependency of a custom flight mode - with the possible exception of core things such as the uorb interface itself, and from now on things like uorb messages.

If that is what is alluded to here, then how about?


```suggestion
- Easier to maintain as the integration API is small, well defined, and stable.
- Porting custom PX4 modes on the flight controller between PX4 versions can be much harder, as often flight modes use interfaces that are considered internal, and allowed to change. 
```

mrpollo commented on en/concept/flight_modes.md:
```diff
@@ -1,111 +1,215 @@
+
+- Easier to implement as there is no need to deal with low-level embedded constraints and requirements (such as restricted stack sizes).
+- Easier to maintain as the integration API is small, well defined, and stable (porting custom PX4 modes on the flight controller between PX4 versions is much harder).
+- Crashing the app does not crash the vehicle.
```
While this is technically true, a poorly written mode can still crash the vehicle, maybe rewrite to say something like "Crashing the process results in a fallback to internal flight modes" or something similar

hamishwillee replied:
```suggestion
- Process termination of a ROS 2 mode results in a fallback to an internal flight mode (while termination of an internal mode may well crash the vehicle).
```

mrpollo commented on en/concept/flight_modes.md:
```diff
@@ -1,111 +1,215 @@
+- High-level functionality is available, including a better-feature programming environment, and many useful Linux and ROS libraries.
+- More available compute to do more advanced processing (e.g. computer vision).
+
+Note that there are some negatives to keep in mind:
```
I think we should open an issue where we track these three bullet points and link to this instead of listing them here where they will likely get stale over time

hamishwillee replied:
@mrpollo I propose you do that - the point about the control interface being experimental is relevant here but probably wouldn't go in the same issue. We do need to talk to @bkueng about what is needed to get it to the point it is considered "stable".

I did create this one just for the CI:
- https://github.com/PX4/PX4-Autopilot/issues/24363

mrpollo commented on en/concept/flight_modes.md:
```diff
@@ -1,111 +1,215 @@
+
+[PX4 ROS 2 Control Interface](../ros2/px4_ros2_control_interface.md) documents how to create flight modes running on a companion computer.
 
+## PX Flight Modes (FC)
```
```suggestion
## PX4 Flight Modes (FC)
```

mrpollo reviewed (commented):
I hope this helps

hamishwillee commented:
@mrpollo @sfuhrer OK, updated to use external/internal. I explained above the potential issue with this, but I think we can live with it for now. Also removed the bits that were worrying you @mrpollo.

What worries me about this is that you still couldn't create a flight mode with this information - i.e. we need an answer to https://github.com/PX4/PX4-user_guide/pull/3577#discussion_r1953934066

hamishwillee commented:
@mrpollo @sfuhrer @MaEtUgR 
- I removed the information on how to create PX4 Internal modes other than the mode requirement section (which is accurate)
- I moved the standard modes integration docs into a standalone MAVLink doc, because you might want to know about that without knowing about the rest of this stuff - and linked

This still can't be used for creating internal modes, but it is not "incorrect" and it is more useful than what it replaces.

Merging for now, but I still need the extra info.
